pkg/controller/direct/apigee/instanceattachment_controller.go

```diff
@@ -0,0 +1,297 @@
+
+	// Resolve references
+	if err := ResolveApigeeInstanceAttachmentRefs(ctx, a.k8sClient, desired); err != nil {
+		return err
+	}
+	// Convert to proto
+	resource := ApigeeInstanceAttachmentSpec_ToAPI(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	// HACK: Environment required by GCP API is name-only, not fully-qualified ID.
+	// So, we fix this by editing the environment to be name-only.
+	environmentID := &apigeev1beta1.ApigeeEnvironmentIdentity{}
+	if err := environmentID.FromExternal(resource.Environment); err != nil {
+		return err
+	}
+	resource.Environment = environmentID.ResourceID
+
+	op, err := a.attachmentsClient.Create(a.id.ParentID.String(), resource).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("creating ApigeeInstanceAttachment %s: %w", a.desired.Name, err)
+	}
+	if err := WaitForApigeeOp(ctx, a.operationsClient, op); err != nil {
+		return fmt.Errorf("ApigeeInstanceAttachment %s waiting creation: %w", a.desired.Name, err)
+	}
+
+	// Get response from completed operation and unmarshal it so that we can determine
+	// the server-generated ID for the newly-created instance attachment.
+	result, err := a.operationsClient.Get(op.Name).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("getting create ApigeeInstanceAttachment %s operation: %w", a.desired.Name, err)
+	}
+	var resultResponse api.GoogleCloudApigeeV1InstanceAttachment
+	if err := json.Unmarshal(result.Response, &resultResponse); err != nil {
+		return fmt.Errorf("unmarshalling create ApigeeInstanceAttachment %s operation response: %w", a.desired.Name, err)
+	}
+
+	// Update identity to include server-generated ID.
+	a.id.ResourceID = resultResponse.Name
+
+	// Get the newly-created instance attachment.
+	created, err := a.attachmentsClient.Get(a.id.String()).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("getting created ApigeeInstanceAttachment %s: %w", a.desired.Name, err)
+	}
+
+	log.V(2).Info("successfully created ApigeeInstanceAttachment", "name", a.desired.Name, "id", a.id)
+
+	status := &krm.ApigeeInstanceAttachmentStatus{}
+	status.ObservedState = ApigeeInstanceAttachmentObservedState_FromAPI(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *ApigeeInstanceAttachmentAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating ApigeeInstanceAttachment", "id", a.id)
+	mapCtx := &direct.MapContext{}
+
+	// There are no fields in the GCP ApigeeInstanceAttachment API that can be updated.
+	// So, we will only update the KRM status.
+	status := &krm.ApigeeInstanceAttachmentStatus{}
+	status.ObservedState = ApigeeInstanceAttachmentObservedState_FromAPI(mapCtx, a.actual)
```

Comment:
Some future improvements: use etag or a snapshot to decide if we can skip status update.